src/software/world/ball.cpp

```diff
@@ -55,11 +57,30 @@ Vector Ball::acceleration() const
 
 BallState Ball::estimateFutureState(const Duration &duration_in_future) const
 {
+    Vector future_velocity = calculateFutureVelocity(current_state_.velocity(),
+                                                     acceleration_, duration_in_future);
+
+    // since acceleration is due to friction, final velocity cannot be in the opposite
+    // direction to initial velocity. if that happens, we find a new velocity using a more
+    // realistic Duration
+    Duration effective_duration(duration_in_future);
+
+
+    if (acceleration_.length() > 0 &&
+        almostEqual(future_velocity.rotate(Angle::half()).orientation().toRadians(),
+                    current_state_.velocity().orientation().toRadians(), FIXED_EPSILON,
+                    ULPS_EPSILON_TEN))
```

Comment:
this is pretty confusing, can you explain and add a comment

Reply:
its checking if the final velocity (after friction has been applied for the whole duration) is in the opposite direction to the velocity the ball is currently in. We check this because friction will only be applied until the ball completely stops. So if the final velocity is in the opposite direction we need to use a different duration to find position. 

there is a comment above the if statement with explanation, i moved this check into its own variable to make it clearer